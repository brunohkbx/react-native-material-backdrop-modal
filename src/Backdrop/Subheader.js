import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import Animated from 'react-native-reanimated';
import Touchable from 'react-native-platform-touchable';
import PropTypes from 'prop-types';

const AnimatedTouchable = Animated.createAnimatedComponent(Touchable);

const Subheader = ({ disabled, onPress, icon, iconContainerStyle, title }) => {
  return (
    <Touchable
      disabled={disabled}
      onPress={onPress}
      background={Touchable.Ripple('rgba(0, 0, 0, 0.2)', true)}
      style={{ height: 48 }}
    >
      <View style={styles.headerContainer}>
        <Text style={styles.title}>{title}</Text>
        {icon && (
          <AnimatedTouchable
            onPress={onPress}
            centered
            style={[styles.icon, iconContainerStyle]}
            background={Touchable.Ripple('rgba(0, 0, 0, 0.2)', true)}
            testID="icon"
          >
            {icon}
          </AnimatedTouchable>
        )}
      </View>
    </Touchable>
  );
};

Subheader.propTypes = {
  disabled: PropTypes.bool,
  onPress: PropTypes.func.isRequired,
  icon: PropTypes.element,
  iconContainerStyle: PropTypes.object,
  title: PropTypes.string.isRequired,
};

Subheader.defaultProps = {
  disabled: false,
  icon: null,
  iconContainerStyle: {},
};

const styles = StyleSheet.create({
  headerContainer: {
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: 'rgba(0, 0, 0, 0.12)',
    marginHorizontal: 16,
    paddingVertical: 12,
  },
  title: {
    fontSize: 16,
    lineHeight: 24,
    color: 'rgba(0, 0, 0, 0.87)',
    fontFamily: 'sans-serif',
  },
  icon: {
    position: 'absolute',
    top: 12,
    right: 16,
    width: 24,
    height: 24,
  },
});

export default Subheader;
