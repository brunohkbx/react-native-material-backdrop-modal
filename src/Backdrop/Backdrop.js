import React, { useRef } from 'react';
import { StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import { useLayout } from 'react-native-hooks';
import useUpdateEffect from 'react-use/lib/useUpdateEffect';
import Animated, { Easing } from 'react-native-reanimated';
import Subheader from './Subheader';

const SUBHEADER_HEIGHT = 48;

const {
  Clock,
  Value,
  block,
  cond,
  clockRunning,
  set,
  startClock,
  timing,
  stopClock,
  interpolate,
  concat,
} = Animated;

const runTiming = ({ toValue, position = new Value(0), duration = 300 }) => {
  const clock = new Clock();

  const state = {
    finished: new Value(0),
    frameTime: new Value(0),
    position,
    time: new Value(0),
  };

  const config = { toValue, duration, easing: Easing.inOut(Easing.ease) };

  return block([
    cond(clockRunning(clock), 0, [
      set(state.finished, 0),
      set(state.frameTime, 0),
      set(state.time, 0),
      set(config.toValue, toValue),
      startClock(clock),
    ]),
    timing(clock, state, config),
    cond(state.finished, stopClock(clock)),
    state.position,
  ]);
};

const Backdrop = ({ children, focused, onFocus, title, icon }) => {
  const { onLayout: onLayoutOverlay, height: overlayHeight } = useLayout();
  const revealedPosition = overlayHeight - SUBHEADER_HEIGHT;

  const baseValue = useRef(
    focused ? new Value(revealedPosition) : new Value(0)
  );

  useUpdateEffect(() => {
    baseValue.current = runTiming({
      toValue: new Value(focused ? revealedPosition : 0),
      position: new Value(focused ? 0 : revealedPosition),
    });
  }, [focused]);

  const opacityIcon = interpolate(baseValue.current, {
    inputRange: [-SUBHEADER_HEIGHT, 0, Math.abs(revealedPosition)],
    outputRange: [0, 1, 0],
  });

  const childrenHeight = concat(
    interpolate(baseValue.current, {
      inputRange: [-SUBHEADER_HEIGHT, 0, Math.abs(revealedPosition)],
      outputRange: [100, 0, 100],
    }),
    '%'
  );

  return (
    <View style={[styles.overlay]} onLayout={onLayoutOverlay}>
      <View style={[styles.backdrop]}>
        <Subheader
          disabled={focused}
          onPress={onFocus}
          icon={icon}
          iconContainerStyle={{ opacity: opacityIcon }}
          title={title}
          testID="subheader"
        />
        <Animated.View style={{ height: childrenHeight }}>
          <View style={styles.children}>{children}</View>
        </Animated.View>
      </View>
    </View>
  );
};

Backdrop.propTypes = {
  children: PropTypes.node,
  focused: PropTypes.bool,
  onFocus: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  icon: PropTypes.element,
};

Backdrop.defaultProps = {
  children: null,
  focused: true,
  icon: null,
};

const styles = StyleSheet.create({
  overlay: {
    ...StyleSheet.absoluteFill,
    justifyContent: 'flex-end',
  },
  backdrop: {
    maxHeight: '100%',
    backgroundColor: '#FFF',
    borderTopLeftRadius: 16,
    borderTopRightRadius: 16,
  },
  children: {
    marginHorizontal: 16,
    paddingVertical: 12,
  },
});

export default Backdrop;
