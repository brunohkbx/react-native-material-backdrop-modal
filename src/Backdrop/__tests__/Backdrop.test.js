import React from 'react';
import { View, Text } from 'react-native';
import { fireEvent, render } from 'react-native-testing-library';
import Backdrop from '../Backdrop';

jest.mock('react-native-reanimated', () =>
  // eslint-disable-next-line global-require
  require('react-native-reanimated/mock')
);

describe('Backdrop', () => {
  const setup = (propOverrides = {}) => {
    const defaultProps = {
      onFocus: jest.fn(),
      title: 'Movies',
      ...propOverrides,
    };

    const element = (
      <Backdrop {...defaultProps}>
        <View>
          <Text>Movie 1</Text>
        </View>
      </Backdrop>
    );

    const wrapper = render(element);

    return { element, wrapper };
  };

  beforeAll(() => jest.useFakeTimers());

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  describe('onFocus', () => {
    it('calls onFocus when pressing on Subheader', () => {
      const mockOnFocus = jest.fn();
      const {
        wrapper: { getByTestId },
      } = setup({ focused: false, onFocus: mockOnFocus });

      fireEvent.press(getByTestId('subheader'));

      expect(mockOnFocus).toHaveBeenCalled();
    });

    it("calls onFocus when pressing on Subheader's icon", () => {
      const mockOnFocus = jest.fn();
      const {
        wrapper: { getByTestId },
      } = setup({ focused: false, onFocus: mockOnFocus, icon: <View /> });

      fireEvent.press(getByTestId('icon'));

      expect(mockOnFocus).toHaveBeenCalled();
    });
  });
});
