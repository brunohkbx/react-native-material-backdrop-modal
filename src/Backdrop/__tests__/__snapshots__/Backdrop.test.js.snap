// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop renders properly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "bottom": 0,
        "justifyContent": "flex-end",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFF",
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "maxHeight": "100%",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 48,
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "borderBottomColor": "rgba(0, 0, 0, 0.12)",
            "borderBottomWidth": 0.5,
            "marginHorizontal": 16,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "sans-serif",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          Movies
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "height": undefined,
        }
      }
    >
      <View
        style={
          Object {
            "marginHorizontal": 16,
            "paddingVertical": 12,
          }
        }
      >
        <View>
          <Text>
            Movie 1
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
